File: services/ocaml-oxcaml/lib/protocol.ml
open! Core

let frame_bytes = 640

module Start_message = struct
  type t = { type_ : string }

  let of_yojson (json : Yojson.Safe.t) : (t, string) Result.t =
    match json with
    | `Assoc fields ->
      (match List.Assoc.find fields ~equal:String.equal "type" with
       | Some (`String value) when List.length fields = 1 -> Ok { type_ = value }
       | Some _ -> Error "Start_message: rejected unknown fields"
       | None -> Error "Start_message: missing required \"type\" field")
    | _ -> Error "Start_message: expected JSON object"
  ;;
end

module Infer_response = struct
  type t =
    { rms : float
    ; zero_crossings : int
    ; checksum : int
    ; samples : int
    ; transcript : string
    ; audio_bytes : int
    }

  (* Fields the inference server returns. Anything else -> error. *)
  let known_keys =
    String.Set.of_list
      [ "rms"; "zero_crossings"; "checksum"; "samples"; "transcript"; "audio_bytes" ]
  ;;

  let read_float = function
    | `Float f -> Ok f
    | `Int i -> Ok (Float.of_int i)
    | _ -> Error "expected number"
  ;;

  let read_int = function
    | `Int i -> Ok i
    | `Intlit s ->
      (match Int.of_string_opt s with
       | Some i -> Ok i
       | None -> Error "Intlit out of range")
    | _ -> Error "expected integer"
  ;;

  let read_string = function
    | `String s -> Ok s
    | _ -> Error "expected string"
  ;;

  let of_yojson (json : Yojson.Safe.t) : (t, string) Result.t =
    match json with
    | `Assoc fields ->
      let unknown =
        List.filter_map fields ~f:(fun (k, _) ->
          if Set.mem known_keys k then None else Some k)
      in
      if not (List.is_empty unknown)
      then
        Error
          (sprintf
             "Infer_response: rejected unknown fields: %s"
             (String.concat ~sep:"," unknown))
      else
        let open Result.Let_syntax in
        let find key reader =
          match List.Assoc.find fields ~equal:String.equal key with
          | Some v -> reader v
          | None -> Error (sprintf "Infer_response: missing %s" key)
        in
        let%bind rms = find "rms" read_float in
        let%bind zero_crossings = find "zero_crossings" read_int in
        let%bind checksum = find "checksum" read_int in
        let%bind samples = find "samples" read_int in
        let%bind transcript = find "transcript" read_string in
        let%bind audio_bytes = find "audio_bytes" read_int in
        Ok { rms; zero_crossings; checksum; samples; transcript; audio_bytes }
    | _ -> Error "Infer_response: expected JSON object"
  ;;
end

module Partial = struct
  type t =
    { type_ : string
    ; oldest_frame_seq : int
    ; newest_frame_seq : int
    ; frames : int
    ; rms : float
    ; zero_crossings : int
    ; checksum : int
    ; samples : int
    ; transcript : string
    ; audio_bytes : int
    ; cpu_passes : int
    ; model_delay_ms : int
    ; flush_lateness_ms : float
    ; inflight_model_jobs : int
    }

  let to_yojson (t : t) : Yojson.Safe.t =
    `Assoc
      [ "type", `String t.type_
      ; "oldest_frame_seq", `Int t.oldest_frame_seq
      ; "newest_frame_seq", `Int t.newest_frame_seq
      ; "frames", `Int t.frames
      ; "rms", `Float t.rms
      ; "zero_crossings", `Int t.zero_crossings
      ; "checksum", `Int t.checksum
      ; "samples", `Int t.samples
      ; "transcript", `String t.transcript
      ; "audio_bytes", `Int t.audio_bytes
      ; "cpu_passes", `Int t.cpu_passes
      ; "model_delay_ms", `Int t.model_delay_ms
      ; "flush_lateness_ms", `Float t.flush_lateness_ms
      ; "inflight_model_jobs", `Int t.inflight_model_jobs
      ]
  ;;

  let to_string t = Yojson.Safe.to_string (to_yojson t)
end

module Error = struct
  type t =
    { type_ : string
    ; stage : string
    ; kind : string
    ; message : string
    ; oldest_frame_seq : int
    ; newest_frame_seq : int
    ; frames : int
    ; audio_bytes : int
    ; oldest_age_ms : float
    ; newest_age_ms : float
    ; flush_lateness_ms : float
    ; inference_elapsed_ms : float option
    ; inflight_gateway_batches : int
    ; gateway_buffer_frames : int
    ; inference_status : int option
    ; retryable : bool
    }

  let float_or_null = function
    | None -> `Null
    | Some f -> `Float f
  ;;

  let int_or_null = function
    | None -> `Null
    | Some i -> `Int i
  ;;

  let to_yojson (t : t) : Yojson.Safe.t =
    `Assoc
      [ "type", `String t.type_
      ; "stage", `String t.stage
      ; "kind", `String t.kind
      ; "message", `String t.message
      ; "oldest_frame_seq", `Int t.oldest_frame_seq
      ; "newest_frame_seq", `Int t.newest_frame_seq
      ; "frames", `Int t.frames
      ; "audio_bytes", `Int t.audio_bytes
      ; "oldest_age_ms", `Float t.oldest_age_ms
      ; "newest_age_ms", `Float t.newest_age_ms
      ; "flush_lateness_ms", `Float t.flush_lateness_ms
      ; "inference_elapsed_ms", float_or_null t.inference_elapsed_ms
      ; "inflight_gateway_batches", `Int t.inflight_gateway_batches
      ; "gateway_buffer_frames", `Int t.gateway_buffer_frames
      ; "inference_status", int_or_null t.inference_status
      ; "retryable", `Bool t.retryable
      ]
  ;;

  let to_string t = Yojson.Safe.to_string (to_yojson t)
end

File: services/rust-sync/src/protocol.rs
//! Wire types for the shared `/ws/stt` protocol.
//!
//! Implemented to the black-box contract (CLAUDE.md "Shared protocol
//! contract" + `loadgen/rust/src/protocol.rs`), not by copying another
//! gateway's internals. Both boundaries are strict: unknown fields are
//! rejected, mirroring serde `deny_unknown_fields` / Pydantic
//! `extra="forbid"` used by the other runtimes.

use serde::{Deserialize, Serialize};

/// First client message. Must parse as `{"type":"start"}`.
///
/// `deny_unknown_fields` is intentionally *not* set: serde silently
/// ignores it on internally-tagged enums, and the canonical contract type
/// (`loadgen/rust/src/protocol.rs::StartMessage`) is the same shape with
/// the same quirk. Matching the reference's leniency on extra fields is a
/// black-box-substitutability requirement, not an oversight — the
/// `invalid start` case the conformance harness exercises is non-JSON /
/// wrong-`type`, both of which still fail to parse here.
#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StartMessage {
    Start,
}

/// Inference server `/infer` response envelope.
#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct InferResponse {
    pub rms: f64,
    pub zero_crossings: u64,
    pub checksum: u32,
    pub samples: u64,
    pub transcript: String,
    pub audio_bytes: u64,
}

/// Server → client success frame. Field set and `type` tag are fixed by
/// the contract; the loadgen computes latency curves from
/// `oldest_frame_seq`/`newest_frame_seq`/`flush_lateness_ms`.
#[derive(Debug, Serialize)]
pub struct PartialMessage {
    #[serde(rename = "type")]
    pub type_tag: &'static str,
    pub oldest_frame_seq: u64,
    pub newest_frame_seq: u64,
    pub frames: u64,
    pub rms: f64,
    pub zero_crossings: u64,
    pub checksum: u32,
    pub samples: u64,
    pub transcript: String,
    pub audio_bytes: u64,
    pub cpu_passes: u32,
    pub model_delay_ms: u64,
    pub flush_lateness_ms: f64,
    pub inflight_model_jobs: u64,
}

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorStage {
    InferenceRequest,
    InferenceResponseParse,
}

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ErrorKind {
    Timeout,
    // serde's snake_case strips the underscore before digits
    // (`Http5xx` → `"http5xx"`); rename to the wire spelling the other
    // runtimes use.
    #[serde(rename = "http_5xx")]
    Http5xx,
    #[serde(rename = "http_429")]
    Http429,
    ConnectionReset,
    ParseError,
}

/// Server → client diagnostic frame emitted when a flush batch can't be
/// served. The connection stays open; later flushes resume normally.
#[derive(Debug, Serialize)]
pub struct ErrorMessage {
    #[serde(rename = "type")]
    pub type_tag: &'static str,
    pub stage: ErrorStage,
    pub kind: ErrorKind,
    pub message: String,
    pub oldest_frame_seq: u64,
    pub newest_frame_seq: u64,
    pub frames: u64,
    pub audio_bytes: u64,
    pub oldest_age_ms: f64,
    pub newest_age_ms: f64,
    pub flush_lateness_ms: f64,
    pub inference_elapsed_ms: Option<f64>,
    pub inflight_gateway_batches: u64,
    pub gateway_buffer_frames: u64,
    pub inference_status: Option<u16>,
    pub retryable: bool,
}

pub const PARTIAL_TYPE: &str = "partial";
pub const ERROR_TYPE: &str = "error";

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn start_accepts_exact_contract_shape() {
        assert!(serde_json::from_str::<StartMessage>(r#"{"type":"start"}"#).is_ok());
    }

    #[test]
    fn start_rejects_non_json_and_wrong_type() {
        // The two shapes the `invalid start` conformance case actually
        // exercises. Extra fields are accepted on purpose — see the
        // `StartMessage` doc comment (matches the reference contract).
        assert!(serde_json::from_str::<StartMessage>("not json").is_err());
        assert!(serde_json::from_str::<StartMessage>(r#"{"type":"begin"}"#).is_err());
        assert!(serde_json::from_str::<StartMessage>(r#"{"type":"start","x":1}"#).is_ok());
    }

    #[test]
    fn infer_response_rejects_unknown_fields() {
        let ok = r#"{"rms":1.0,"zero_crossings":2,"checksum":3,"samples":4,"transcript":"now","audio_bytes":640}"#;
        assert!(serde_json::from_str::<InferResponse>(ok).is_ok());
        let extra = r#"{"rms":1.0,"zero_crossings":2,"checksum":3,"samples":4,"transcript":"now","audio_bytes":640,"x":1}"#;
        assert!(serde_json::from_str::<InferResponse>(extra).is_err());
    }

    #[test]
    fn partial_serializes_with_type_tag_and_full_field_set() {
        let partial = PartialMessage {
            type_tag: PARTIAL_TYPE,
            oldest_frame_seq: 1,
            newest_frame_seq: 50,
            frames: 50,
            rms: 2425.5,
            zero_crossings: 7,
            checksum: 1_745_339_397,
            samples: 1280,
            transcript: "now".to_owned(),
            audio_bytes: 32000,
            cpu_passes: 4,
            model_delay_ms: 75,
            flush_lateness_ms: 0.0,
            inflight_model_jobs: 0,
        };
        let value: serde_json::Value =
            serde_json::from_slice(&serde_json::to_vec(&partial).unwrap()).unwrap();
        assert_eq!(value["type"], "partial");
        for field in [
            "oldest_frame_seq",
            "newest_frame_seq",
            "frames",
            "rms",
            "zero_crossings",
            "checksum",
            "samples",
            "transcript",
            "audio_bytes",
            "cpu_passes",
            "model_delay_ms",
            "flush_lateness_ms",
            "inflight_model_jobs",
        ] {
            assert!(value.get(field).is_some(), "missing {field}");
        }
    }

    #[test]
    fn error_kind_uses_wire_spelling() {
        let json = serde_json::to_string(&ErrorKind::Http5xx).unwrap();
        assert_eq!(json, r#""http_5xx""#);
    }
}
